src/main/java/fr/xephi/authme/listener/PlayerListener.java

```diff
@@ -229,175 +225,162 @@ public void onJoinMessage(PlayerJoinEvent event) {
+            || permissionsManager.hasPermission(player, PlayerStatePermission.ALLOW_CHAT_BEFORE_LOGIN);
+        if (mayPlayerSendChat) {
+            removeUnauthorizedRecipients(event);
+        } else {
+            event.setCancelled(true);
+            messages.send(player, MessageKey.DENIED_CHAT);
+        }
+    }
 
-        if (validationService.isUnrestricted(name)) {
+    @EventHandler(ignoreCancelled = true, priority = EventPriority.LOWEST)
+    public void onPlayerCommandPreprocess(PlayerCommandPreprocessEvent event) {
+        String cmd = event.getMessage().split(" ")[0].toLowerCase();
+        if (settings.getProperty(HooksSettings.USE_ESSENTIALS_MOTD) && "/motd".equals(cmd)) {
             return;
         }
-
-        // Keep pre-UUID compatibility
-        try {
-            permissionsManager.loadUserData(event.getUniqueId());
-        } catch (PermissionLoadUserException e) {
-            logger.logException("Unable to load the permission data of user " + name, e);
+        if (settings.getProperty(RestrictionSettings.ALLOW_COMMANDS).contains(cmd)) {
+            return;
         }
-
-        // getAddress() sometimes returning null if not yet resolved
-        // skip it and let PlayerLoginEvent to handle it
-        if (event.getAddress() == null) {
-            unresolvedPlayerHostname.add(event.getName());
+        final Player player = event.getPlayer();
+        if (!quickCommandsProtectionManager.isAllowed(player.getName())) {
+            event.setCancelled(true);
+            player.kickPlayer(messages.retrieveSingle(player, MessageKey.QUICK_COMMAND_PROTECTION_KICK));
             return;
-        } else {
-            unresolvedPlayerHostname.remove(event.getName());
         }
-
-        try {
-            runOnJoinChecks(JoiningPlayer.fromName(name), event.getAddress().getHostAddress());
-        } catch (FailedVerificationException e) {
-            event.setKickMessage(messages.retrieveSingle(name, e.getReason(), e.getArgs()));
-            event.setLoginResult(AsyncPlayerPreLoginEvent.Result.KICK_OTHER);
+        if (listenerService.shouldCancelEvent(player)) {
+            event.setCancelled(true);
+            messages.send(player, MessageKey.DENIED_COMMAND);
         }
     }
 
-    //Note: We can't teleport the player in the PlayerLoginEvent listener
-    //as the new player location will be reverted by the server.
-
-    @EventHandler(priority = EventPriority.LOW)
-    public void onPlayerLogin(PlayerLoginEvent event) {
-        final Player player = event.getPlayer();
-        final String name = player.getName();
+    /*
+     * Movement events
+     */
 
-        if (validationService.isUnrestricted(name)) {
+    @EventHandler(ignoreCancelled = true, priority = EventPriority.HIGHEST)
+    public void onPlayerMove(PlayerMoveEvent event) {
+        if (settings.getProperty(ALLOW_UNAUTHED_MOVEMENT) && settings.getProperty(ALLOWED_MOVEMENT_RADIUS) <= 0) {
```

Comment:
On L353 the same `ALLOW_UNAUTHED_MOVEMENT` is checked if it's false; I would merge these two conditions at the start of this method like so:
```java
if (!settings.getProperty(ALLOW_UNAUTHED_MOVEMENT)) { // taken from L353 onwards
            // "cancel" the event
            event.setTo(event.getFrom());
            return;
} else if (settings.getProperty(ALLOWED_MOVEMENT_RADIUS) <= 0) {
  return; // unrestricted
}
```

Reply:
Not applicable, the movement logic wouldn't make sense.

Reply:
How does moving a check up in the method affect the movement logic?

Reply:
We want to allow movement on the Y axis

Reply:
Ah sorry, you're right :relaxed: